File: android/xray/src/main/java/com/applicaster/xray/routing/Mapper.java
package com.applicaster.xray.routing;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.applicaster.xray.Logger;
import com.applicaster.xray.sinks.ISink;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

/*
  Maps loggers to sinks using rules. Not thread safe
 */
public class Mapper {

    // sinks Set is be backed up by original map owned by Core
    private final Set<String> sinks;
    // map Logger base name -> (sink name -> sink filter)
    private final Map<String, HashMap<String, ISinkFilter>> loggerMapping = new HashMap<>();

    public Mapper(HashMap<String, ISink> sinks) {
        this.sinks = sinks.keySet();
    }

    public void setFilter(@NonNull final String loggerName,
                          @NonNull final String sinkName,
                          @NonNull final ISinkFilter filter) {
        synchronized (loggerMapping) {
            HashMap<String, ISinkFilter> mappedSinks = loggerMapping.get(loggerName);
            if (null != mappedSinks) {
                mappedSinks.put(sinkName, filter);
                return;
            }
            loggerMapping.put(loggerName, new HashMap<String, ISinkFilter>() {{
                put(sinkName, filter);
            }});
        }
    }

    @Nullable
    private HashMap<String, ISinkFilter> getClosestMapping(@NonNull String logger) {
        if(logger.isEmpty()) {
            return loggerMapping.get(""); // root logger mapping
        }
        // exact match
        HashMap<String, ISinkFilter> mappedSinks = loggerMapping.get(logger);
        if(null != mappedSinks) {
            return mappedSinks;
        }
        // look for any parent mapping
        int i = logger.lastIndexOf(Logger.NameSeparator);
        while(i > 0) {
            String parent = logger.substring(0, i);
            mappedSinks = loggerMapping.get(parent);
            if(null != mappedSinks) {
                return mappedSinks;
            }
            logger = parent;
            i = logger.lastIndexOf(Logger.NameSeparator);
        }
        return loggerMapping.get(""); // root logger mapping
    }

    public Set<String> getMapping(@NonNull String logger,
                                  @NonNull String tag,
                                  int level) {
        HashSet<String> result;
        synchronized (sinks) {
            result = new HashSet<>(this.sinks);
        }
        HashMap<String, ISinkFilter> mappedSinks;
        synchronized (loggerMapping) {
            mappedSinks = getClosestMapping(logger);
        }
        if (mappedSinks == null) {
            return result;
        }
        for (Map.Entry<String, ISinkFilter> mappedSink : mappedSinks.entrySet()) {
            if (null != mappedSink.getValue() && !mappedSink.getValue().accept(logger, tag, level)) {
                result.remove(mappedSink.getKey());
            }
        }
        return result;
    }

    public boolean hasSinks(@NonNull String loggerName,
                            @NonNull String tag,
                            int level) {
        // todo: optimize
        return !getMapping(loggerName, tag, level).isEmpty();
    }
}

File: android/xray-core/src/main/java/com/applicaster/xray/core/routing/Mapper.kt
package com.applicaster.xray.core.routing

import com.applicaster.xray.core.ISink
import com.applicaster.xray.core.Logger

/*
 Maps loggers to sinks using rules. Not thread safe
*/
class Mapper(private val sinks: Map<String, ISink?>) {

    // map Logger subsystem -> (sink identifier -> sink filter)
    private val loggerMapping: MutableMap<String, MutableMap<String, ISinkFilter?>> = mutableMapOf()

    fun setFilter(
        loggerName: String,
        sinkName: String,
        filter: ISinkFilter?
    ) {
        synchronized(loggerMapping) {
            val mappedSinks = loggerMapping[loggerName]
            if (null != mappedSinks) {
                if (null != filter) {
                    mappedSinks[sinkName] = filter
                } else {
                    mappedSinks.remove(sinkName)
                    if (mappedSinks.isEmpty()) {
                        loggerMapping.remove(loggerName)
                    }
                }
                return
            }
            if (null == filter) {
                return
            }
            loggerMapping.put(loggerName, mutableMapOf(sinkName to filter))
        }
    }

    private fun getClosestMapping(logger: String): Map<String, ISinkFilter?>? {
        if (logger.isEmpty()) {
            return loggerMapping[""] // root logger mapping
        }
        // exact match
        var mappedSinks = loggerMapping[logger]
        if (null != mappedSinks) {
            // should not be empty
            return mappedSinks
        }
        // look for any parent mapping
        var i = logger.lastIndexOf(Logger.NameSeparator)
        var parent = logger
        while (i > 0) {
            parent = parent.substring(0, i)
            mappedSinks = loggerMapping[parent]
            if (null != mappedSinks) {
                return mappedSinks
            }
            i = parent.lastIndexOf(Logger.NameSeparator)
        }
        return loggerMapping[""] // root logger mapping
    }

    fun getMapping(
        logger: String,
        tag: String,
        level: Int
    ): Set<String> {
        val result = synchronized(sinks) {
            HashSet(this.sinks.keys)
        }
        val mappedSinks = synchronized(loggerMapping) {
            getClosestMapping(logger)
        }
        if (mappedSinks == null) {
            return result
        }
        for ((key, value) in mappedSinks) {
            if (null != value && !value.accept(logger, tag, level)) {
                result.remove(key)
            }
        }
        return result
    }

    fun hasSinks(
        loggerName: String,
        tag: String,
        level: Int
    ): Boolean {
        // todo: optimize
        return getMapping(loggerName, tag, level).isNotEmpty()
    }

    fun reset() {
        synchronized(loggerMapping) {
            loggerMapping.clear()
        }
    }
}
